contracts/contracts/rem.swap/include/rem.swap/rem.swap.hpp

```diff
@@ -0,0 +1,210 @@
+
+    private:
+        enum class swap_status: uint8_t {
+            INIT = 0,
+            FINISH = 1
+        };
+
+        const asset create_account_fee = {20, symbol("REM", 4)};
+        const string remchain_id = "cf057bbfb72640471fd910bcb67639c22df9f92470936cddc1ade0e2f2e7dc4f";
+        const time_point swap_lifetime = time_point(seconds(15552000)); // 180 days
+
+        struct approval {
+            permission_level level;
+            time_point       time;
+        };
+
+        struct [[eosio::table]] swap_data {
+            uint64_t                key;
+            string                  trx_id;
+            checksum256             swap_hash;
+            time_point              swap_init_time;
+            uint32_t                timepoint;
+            uint8_t                 status;
+
+            std::vector<approval>   provided_approvals;
+
+            uint64_t primary_key()const { return key; }
+            checksum256 by_swap_hash()const { return swap_hash; }
+
+	   	 	EOSLIB_SERIALIZE( swap_data, (key)(trx_id)(swap_hash)(swap_init_time)
+	   	 	                             (timepoint)(status)(provided_approvals) )
+        };
+
+        struct [[eosio::table]] swap_to_erc20 {
+            uint64_t   key;
+            name       user;
+            string     eth_address;
+            time_point swap_init_time;
+            asset      amount;
+
+            uint64_t primary_key()const { return key; }
+            uint64_t by_name()const { return user.value; }
+        };
+
+        struct permission_level_weight {
+            permission_level permission;
+            uint16_t weight;
+        };
+
+        struct key_weight {
+            public_key key;
+            uint16_t weight;
+        };
+
+        struct wait_weight {
+            uint32_t wait_sec;
+            uint16_t weight;
+        };
+
+        struct authority {
+            uint32_t threshold;
+            std::vector<key_weight> keys;
+            std::vector<permission_level_weight> accounts;
+            std::vector<wait_weight> waits;
+        };
+
+        struct newaccount {
+            name creator;
+            name name;
+            authority owner;
+            authority active;
+        };
+
+        typedef multi_index< "swaps"_n, swap_data, indexed_by <"byhash"_n,
+                const_mem_fun< swap_data, checksum256, &swap_data::by_swap_hash >>
+                > swap_index;
+
+        typedef multi_index<"swapstoerc"_n, swap_to_erc20, indexed_by<"byname"_n,
+                const_mem_fun< swap_to_erc20, uint64_t, &swap_to_erc20::by_name >>
+                > erc20_swap_index;
+
+        bool is_block_producer( const name& user ) {
+            std::vector<name> _producers = get_active_producers();
+            return std::find(_producers.begin(), _producers.end(), user) != _producers.end();
+        }
+
+        checksum256 hash(const string& str) {
+            return sha256(const_cast<char*>(str.c_str()), str.size());
```

Comment:
sha256 accepts const char*

Reply:
It's no matter when you use `const_cast` and `const string`.

Reply:
sha256 accepts const char*
c_str method of string returns const char* too
If sha256 accepted char* (not const) then it will make sense, but it accepts const char* too, so I don't see any reason in const_cast here.